Fix dataverses/{identifier}/metadatablocks endpoint to not return fields marked as displayOnCreate=true if there is an input level with include=false

GPortas opened the pull request:
**What this PR does / why we need it**:

The issue caused fields with displayOnCreate=true to be returned for the create dataset form on the SPA when there was an input-level configuration for the collection and field with include=false.

If an input-level configuration with include=false is set, the field should not be returned even if it has displayOnCreate property set to true, as is the case with JSF, where the field is not displayed in the create dataset form.

This PR also refactors the logic of the JsonPrinter class for filtering fields based on possible conditions by using criteria queries instead of post-filtering conditions within the for loop that iterates over the fields of each metadata block.

**Which issue(s) this PR closes**:

- Closes #10741

**Special notes for your reviewer**:

New tests have been added for the problematic scenario

**Suggestions on how to test this**:

Create a collection from JSF, in the metadata blocks section, uncheck the option to inherit the metadata blocks from the parent. In citation, deselect the Notes field, which is a field with displayOnCreate=true in the database.

Call the following endpoint for the new dataverse, with returnDatasetFieldTypes=true and onlyDisplayedOnCreate=true. This is the same call the SPA uses for the create dataset form.

 curl -H "X-Dataverse-key:<API_TOKEN> "$SERVER_URL/api/dataverses/<DATAVERSE_ID>/metadatablocks?returnDatasetFieldTypes=true&onlyDisplayedOnCreate=true"

The notes text, which has include=false as the input level condition, should not be retrieved.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

No

**Is there a release notes update needed for this change?**:

Yes, attached.

sekmiller commented on src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java:
```diff
@@ -851,4 +852,73 @@ public String getFieldLanguage(String languages, String localeCode) {
+        criteriaQuery.where(criteriaBuilder.and(
+                        criteriaBuilder.equal(metadataBlockRoot.get("id"), metadataBlock.getId()),
+                        datasetFieldTypeRoot.in(metadataBlockRoot.get("datasetFieldTypes")),
+                        criteriaBuilder.isTrue(datasetFieldTypeRoot.get("displayOnCreate"))
```
Do we also need to say "or 'required' is True"? In my local DB producer name is required and displayoncreate is false - not sure it that's just because my local DB is messed up.

sekmiller commented on src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java:
```diff
@@ -851,4 +852,73 @@ public String getFieldLanguage(String languages, String localeCode) {
+                requiredAsInputLevelPredicate
+        )
+                : criteriaBuilder.conjunction();
+
```
Looks good. I haven't done much with criteria builder so maybe adding some comments would help.

Thanks

GPortas commented:
@sekmiller I've updated the code to consider the fields required by the installation (not as a dataverse input level). I’ve added logic to handle the case of conditionally required fields (if a subfield is required but the parent field is not, then do not display it on create).

sekmiller reviewed (approved):
Looks good. Thanks for the updates
